docs/logging.md

```diff
@@ -0,0 +1,101 @@
+# Logging
+
+This document outlines the logging standards and best practices for Ramen.
+
+## General Principles
+
+1. **Consistency**: All controllers should follow the same logging conventions.
+1. **Performance Awareness**: Logging should not significantly impact the
+   performance. Be mindful of the logging levels and the amount of data logged.
+
+## Logger Configuration
+
+- Each controller should create its logger instance with a name specific to the
+  controller for easy identification. For example, if the controller is named
+  `ProtectedVolumeReplicationGroupList`, the logger name could be a shortened
+  form of the controller like `pvrgl`.
+
+  ```go
+  logger := ctrl.Log.WithName("pvrgl")
+  ```
+
+- Loggers should be enriched with key contextual information to make the logs
+  more informative. This includes:
+    - The name of the custom resource.
+    - Then namespace of the custom resource, if applicable.
+    - The UID of the custom resource.
+    - The version of the custom resource.
+
+  ```go
+  logger = logger.WithValues("name", req.NamespacedName.Name, "rid", instance.ObjectMeta.UID, "version", instance.ObjectMeta.ResourceVersion)
+  ```
+
+- The context passed to functions should be augmented with the logger instance:
+
+  ```go
+  ctx = ctrl.LoggerInto(ctx, logger)
+  ```
+
+  This will be useful when we call functions that are outside of the controller
+  scope and the instance.log object is not available to them.
+
+## Log Messages
+
+- Log messages should be concise yet informative. Wherever possible, use
+  structured logging to provide context.
+
+- Start and end of significant operations, like reconciliation, should be logged:
+
+  ```go
+  logger.Info("reconcile start")
+  defer logger.Info("reconcile end", "time spent", time.Since(start))
+  ```
+
+- When logging errors, include the context of the error:
+
+  ```go
+  logger.Error(err, "Error description", "key", "value")
+  ```
+
+- Use key-value pairs instead of concatenating the message and data.
+
+    For example, avoid logging like this:
+
+    ```go
+    log.Info(fmt.Sprintf("Requeuing after %v", duration))
```

Comment:
While the suggested method is better, we need to retain the same keys for various entities, else context is lost. A lot of the existing Sprintf based logging has been to avoid this.

I am not stating what is written here is incorrect, but replacing this style with keys should ensure unique and consistent keys for the various values.

Reply:
Good point.